snuba/clickhouse/translator/visitor.py

```diff
@@ -0,0 +1,125 @@
+from copy import deepcopy
+from dataclasses import replace
+from typing import Optional, TypeVar
+
+from snuba.clickhouse.query import Expression as ClickhouseExpression
+from snuba.datasets.plans.translator.visitor import (
+    ExpressionMapper,
+    MappingExpressionTranslator,
+    TranslationRules,
+)
+from snuba.query.expressions import Expression as SnubaExpression
+from snuba.query.expressions import (
+    CurriedFunctionCall,
+    ExpressionVisitor,
+    FunctionCall,
+    Lambda,
+    SubscriptableReference,
+)
+
+
+class ExpressionTranslator(MappingExpressionTranslator[ClickhouseExpression]):
+    """
+    Translates a Snuba expression into a Clickhouse expression.
+
+    As long as the Clickhouse query AST is basically the same as the Snuba one
+    we add a default rule that makes a copy of the original expression.
+    """
+
+    def __init__(
+        self, translation_rules: TranslationRules[ClickhouseExpression]
+    ) -> None:
+        default_rules = TranslationRules(
+            literals=[DefaultSimpleMapper()],
+            columns=[DefaultSimpleMapper()],
+            # TODO: remove the DefaultSubscriptableMapper translation rule as
+            # soon as we finalize the tags translation rule.
+            subscriptables=[DefaultSubscriptableMapper()],
+            functions=[DefaultFunctionMapper()],
+            curried_functions=[DefaultCurriedFunctionMapper()],
+            arguments=[DefaultSimpleMapper()],
+            lambdas=[DefaultLambdaMapper()],
+        )
+        super().__init__(
+            translation_rules=translation_rules, default_rules=default_rules
+        )
+
+
+TSimpleExp = TypeVar("TSimpleExp", bound=SnubaExpression)
+
+
+class DefaultSimpleMapper(ExpressionMapper[TSimpleExp, ClickhouseExpression]):
```

Comment:
These (from here on) are all the defaults that just produce a copy of the expression. For each structured expression, the children are delegated back to the visitor so, for each node, we can apply the appropriate mapper.